File: 240625/방향에 맞춰 최대로 움직이기/max-movements-with-direction.cpp
#include <iostream>
#include <algorithm>
using namespace std;
#define MAX_N 4
#define DIR_NUM 8

int n, ans;
int dx[DIR_NUM+1] = {0, -1, -1, 0, 1, 1, 1, 0, -1};
int dy[DIR_NUM+1] = {0, 0, 1, 1, 1, 0, -1, -1, -1};
int grid[MAX_N][MAX_N];
int direc[MAX_N][MAX_N];

bool InRange(int x, int y) {
    return 0<=x && x<n && 0<=y && y<n;
}

bool CanGo(int x, int y, int num) {
    return InRange(x, y) && grid[x][y] > num;
}

void FindMax(int x, int y, int cnt) {
    ans = max(ans, cnt);

    for(int i=0; i<n; i++) {
        int nx = x+dx[direc[x][y]]*i, ny = y+dy[direc[x][y]]*i;
        if(CanGo(nx, ny, grid[x][y]))
            FindMax(nx, ny, cnt+1);
    }
}

int main() {
    // 여기에 코드를 작성해주세요.
    cin >> n;
    for(int i=0; i<n; i++) {
        for(int j=0; j<n; j++) {
            cin >> grid[i][j];
        }
    }

    for(int i=0; i<n; i++) {
        for(int j=0; j<n; j++) {
            cin >> direc[i][j];
        }
    }

    int r, c;
    cin >> r>>c;

    FindMax(r-1, c-1, 0);

    cout << ans;
    return 0;
}

File: 250703/방향에 맞춰 최대로 움직이기/max-movements-with-direction.py
n = int(input())

num = [
    list(map(int, input().split()))
    for _ in range(n)
]
move_dir = [
    list(map(int, input().split()))
    for _ in range(n)
]
ans = 0


def can_go(x, y, p_num):
    return 0<=x<n and 0<=y<n and num[x][y] > p_num

def find_max(x, y, cnt):
    global ans
    
    ans = max(ans, cnt)

    dxs, dys = [-1, -1, 0, 1, 1, 1, 0, -1], [0, 1, 1, 1, 0, -1, -1, -1]

    d = move_dir[x][y] -1

    for i in range(n):
        nx, ny = x+dxs[d]*i, y+dys[d]*i

        if can_go(nx, ny, num[x][y]):
            find_max(nx, ny, cnt+1)

r, c = tuple(map(int, input().split()))
find_max(r-1, c-1, 0)
print(ans)